frontend/gateway/gateway.go

```diff
@@ -852,6 +852,16 @@ func (lbf *llbBridgeForwarder) StatFile(ctx context.Context, req *pb.StatFileReq
 	return &pb.StatFileResponse{Stat: st}, nil
 }
 
+func (lbf *llbBridgeForwarder) Evaluate(ctx context.Context, req *pb.EvaluateRequest) (*pb.EvaluateResponse, error) {
+	ctx = tracing.ContextWithSpanFromContext(ctx, lbf.callCtx)
+
+	_, err := lbf.getImmutableRef(ctx, req.Ref, "/")
```

Comment:
Not part of your PR, but why does `getImmutableRef()` take a `path` ? 🤯

Reply:
:eyes: seemingly to print a deceiving error message apparently...

```go
func (lbf *llbBridgeForwarder) getImmutableRef(ctx context.Context, id, path string) (cache.ImmutableRef, error) {
	lbf.mu.Lock()
	ref, ok := lbf.refs[id]
	lbf.mu.Unlock()
	if !ok {
		return nil, errors.Errorf("no such ref: %v", id)
	}
	if ref == nil {
		return nil, errors.Wrap(os.ErrNotExist, path)
	}
```

Technically it's possible for a value in `Result.Ref` to be `nil`, but printing `file does not exist: <path>` for that doesn't feel like the right message - `empty ref` feels like a more apt description of what's gone wrong. Definitely an edge case, a sane frontend should never produce `nil` refs like that.

I'll add a commit on top of the stack in this PR to tidy up the method, good catch.